tests/integration/test_dependency_replacement.py

```diff
@@ -0,0 +1,81 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from os import path
+import tarfile
+
+import utils
+
+
+def test_dependency_replacement(test_env, tmpdir):
+    """
+    Check that proper versions of dependencies were used.
+
+    Process:
+    * Send new request to Cachito API to fetch retrodep with another version of dependency package
+    * Download a bundle archive
+
+    Checks:
+    * Check that the state of request is complete
+    * Check that in the response there is a key "replaces" with dict values which was replaced
+    * Check that dir deps/gomod/pkg/mod/cache/download/github.com/pkg/errors/@v/… contains
+        only the required version
+    * Check that app/go.mod file has replace directive for the specified package
+    """
+    dependency_replacements = test_env["dep_replacement"]["dependency_replacements"]
+    client = utils.Client(test_env["api_url"], test_env["api_auth_type"])
+    response_created_req = client.create_new_request(
+        payload={
+            "repo": test_env["package"]["repo"],
+            "ref": test_env["package"]["ref"],
+            "pkg_managers": test_env["package"]["pkg_managers"],
+            "dependency_replacements": dependency_replacements,
+        },
+    )
+    response = client.wait_for_complete_request(response_created_req)
+    assert response.data["state"] == "complete"
+
+    names_replaced_dependencies = {
+        i["replaces"]["name"] for i in response.data["dependencies"] if i["replaces"] is not None
+    }
+    supposed_replaced_dependencies = set([i["name"] for i in dependency_replacements])
+    assert names_replaced_dependencies == supposed_replaced_dependencies
+
+    bundle_dir_name = tmpdir.join(f"download_{str(response.id)}")
+    bundle_archive = tmpdir.join(f"download_{str(response.id)}.tar.gz")
+    resp = client.download_bundle(response.id, bundle_archive)
+    assert resp.status == 200
+    assert tarfile.is_tarfile(bundle_archive)
+
+    with tarfile.open(bundle_archive, "r:gz") as tar:
+        tar.extractall(bundle_dir_name)
+
+    for dependency in dependency_replacements:
+        dep_name = utils.escape_path_go(dependency["name"])
+        dependency_version_file = path.join(
+            bundle_dir_name,
+            "deps",
+            "gomod",
+            "pkg",
+            "mod",
+            "cache",
+            "download",
+            dep_name,
+            "@v",
+            "list",
+        )
+
+        assert path.exists(dependency_version_file)
+        with open(dependency_version_file, "r") as file:
```

Comment:
Could you move this `open` context manager to above the for loop and store the value of `lines`? Then you don't need to keep opening and reading the file for each replacement.

Reply:
It's opening different file each time (checks the version in all replaced dependencies).
For example: `deps/gomod/pkg/mod/cache/download/github.com/pkg/errors/@v/list`

Reply:
For some dependencies, there are more versions listed in the file:
```
deps/gomod/pkg/mod/cache/download/gopkg.in/check.v1/@v/list
v0.0.0-20161208181325-20d25e280405
v1.0.0-20180628173108-788fd7840127
```